### PR TITLE
feat(AUCTO 3 & 6): Add Login and Signup API

### PR DESCRIPTION
**Pull request by EkamSinghPandher**
Jira: 
[Aucto 3](https://auctomarket.atlassian.net/jira/software/c/projects/AUCTO/boards/2?modal=detail&selectedIssue=AUCTO-3)
[Aucto 6](https://auctomarket.atlassian.net/jira/software/c/projects/AUCTO/boards/2?modal=detail&selectedIssue=AUCTO-6)

### Synopsis

Added Sign Up API and Tests
Added Login API and Tests

### Changes

- [x] Add API
- [x] Add Unit Tests

### CODE REVIEW

File: cmd/web/handlers.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+/*
+Handles error and API response for the Login API for buyers
+*/
+func handleBuyerLogin(c *gin.Context) {
+	var loginData data.LoginData
+	err := c.ShouldBindJSON(&loginData)
+
+	if err != nil {
+		c.JSON(http.StatusBadRequest, gin.H{"message": "Bad request data"})
```
Comment (phatpham9) [inline cmd/web/handlers.go:19]: should we have a more meaningful error message e.g. `Invalid email of password`?

File: cmd/web/handlers.go [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+	}
+
+	signUpResponse, err := buyer.BuyerSignUp(db, signUpData)
+
+	if err != nil {
+		c.JSON(http.StatusBadRequest, gin.H{"message": err.Error()})
+		return
+	}
+
+	c.JSON(http.StatusOK, &signUpResponse)
```
Comment (phatpham9) [inline cmd/web/handlers.go:52]: the status code of a successful resource creation should be `201 Created`

File: api/buyer/buyer.go [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+	var response data.LoginResponseData
+	var hashedPwd string
+	buyerExists, err := doesBuyerEmailExist(db, loginData.Email)
+
+	if err != nil {
+		return response, utils.LogError("Error in checking if buyer email exists", err)
+	}
+
+	if !buyerExists {
+		return response, errors.New("Incorrect username or password")
```
Comment (phatpham9) [inline api/buyer/buyer.go:25]: should the message say `The given email is being used`?

File: api/buyer/buyer.go [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+/*
+Logic for Buyer signup, adds the new user to the database if the email is not already in use.
+Returned response is similar to as if the user logged in.
+*/
+func BuyerSignUp(db *sql.DB, signupData data.SignUpData) (data.LoginResponseData, error) {
+	var response data.LoginResponseData
+
+	buyerExists, err := doesBuyerEmailExist(db, signupData.Email)
+	if err != nil {
+		return response, utils.LogError("Error in checking if buyer email exists", err)
```
Comment (phatpham9) [inline api/buyer/buyer.go:51]: should the message be `The email is being used by another user`?

File: cmd/web/handlers.go (hunk 2 above)
Comment (EkamSinghPandher) [reply cmd/web/handlers.go:52]: Ah ok yeah my bad will change this

File: cmd/web/handlers.go (hunk 1 above)
Comment (EkamSinghPandher) [reply cmd/web/handlers.go]: I dont want to reveal too much in terms of what the error is to the client but will log the more verbose error in the backend.

Comment (phatpham9) [reply cmd/web/handlers.go]: agreed

**Review by phatpham9 — APPROVED** [review]

**Review by phatpham9 — APPROVED** [review]
LGTM

**Pull request merged**